### PR TITLE
Optimize - Requires Core Rebuild(?)

### PR DESCRIPTION
**Pull request by Jiiks**

### CODE REVIEW

File: client/src/index.js [hunk 1]
```diff
@@ -68,5 +83,11 @@ class BetterDiscord {
 if (window.BetterDiscord) {
     Logger.log('main', 'Attempting to inject again?');
 } else {
-    let bdInstance = new BetterDiscord();
+    let instance = null;
+    // eslint-disable-next-line no-inner-declarations
+    function init() {
```
Comment (samogot) [inline client/src/index.js]: why not to use anonymus function here?

File: client/src/modules/patcher.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        if ('string' === typeof mn) return WebpackModules.getModuleByName(mn);
+        if (mn instanceof Array) return WebpackModules.getModuleByProps(mn);
+        return null;
+    }
+    static overrideFn(patch) {
+        return function () {
+            for (const s of patch.supers) {
+                try {
+                    s.fn.apply(this, arguments);
+                } catch (err) { }
```
Comment (samogot) [inline client/src/modules/patcher.js]: silently suppressing is bad! at least add `console.warn` otherwise it would be very hard to find errors

File: client/src/modules/patcher.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            }
+            const retVal = patch.ofn.apply(this, arguments);
+            for (const s of patch.slaves) {
+                try {
+                     s.fn.apply(this, [arguments, { patch, retVal }]);
+                } catch (err) { }
```
Comment (samogot) [inline client/src/modules/patcher.js]: same here

File: client/src/modules/reactcomponents.js [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+ * LICENSE file in the root directory of this source tree. 
+*/
+
+import Patcher from './patcher';
+import WebpackModules from './webpackmodules';
+import DiscordApi from './discordapi';
+import { EmoteModule } from 'builtin';
+import { Reflection } from 'ui';
+
+class Filters {
```
Comment (samogot) [inline client/src/modules/reactcomponents.js]: We already had filter. It was in Utils back then, but now they are in WebpackModules. They should be exported from there and reused here

File: client/src/modules/reactcomponents.js [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+        return (...filters) => (module) => {
+            for (const filter of filters) {
+                if (!filter(module)) return false;
+            }
+            return true;
+        }
+    }
+}
+
+class Helpers {
```
Comment (samogot) [inline client/src/modules/reactcomponents.js]: You intentionally dropped all docs that was in my lib? Please add it here too

File: client/src/modules/patcher.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        const id = patchObject.supers.length + 1;
+        const patch = {
+            id,
+            fn: cb,
+            unpatch: () => patchObject.supers.splice(patchObject.supers.findIndex(slave => slave.id === id), 1)
+        };
+        patchObject.supers.push(patch);
+        return patch;
+    }
+
```
Comment (samogot) [inline client/src/modules/patcher.js]: Oh, crap, I can't read further more. Terrible core style. My original monkey patch wasn't plain simple but at least it was well documented. This code is even more complicated, but also it don't have a single line of comment, method names are not obvious, and variable/parameter names are totally obfuscated. Do you need to pay someone for each extra character you type, or what?

File: client/src/modules/reactcomponents.js [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+            }
+            if (updateOthers) self.forceUpdateOthers();
+        });
+    }
+    */
+    forceUpdateOthers() {
+
+    }
+}
+
```
Comment (samogot) [inline client/src/modules/reactcomponents.js]: Fuck, I was motivated to help with BD development, at least with reviewing of changes to my lib, but what I've seen is terrible. Now don't want to do anything with is. Why did you have to mess everything up? I even can't explain what made me more upset in this change... New strange and in-intuitive interfaces? Lack of documentation and bad codestyle? Usage of module just within this module and moreover, "hiding" global patching of everything inside API module? Maybe it's just my code ownership feeling, and I subconsciously hostile to any changes, because I wrote original code? May be. I also have the feeling that what was done - was done half-assedly and with desire to change something just to be different. Why not to use ready solution? I hope that my feeling is wrong.
If you were my junior at work, I might do more constructive review, commenting every thing unless it became good code - even better than it was in my lib, but I have neither desire nor motivation to do it, because it anyway already merged and I suspect you will anyway ignore everything I wrote.
Hope this project will be fine

File: client/src/modules/patcher.js (hunk 6 above)
Comment (rauenzi) [reply client/src/modules/patcher.js]: I agree this code is quite hard to follow due to the unclear variable names all over the place. I do believe comments/documentation usually follow after getting the functionality down. At least that seems to be the pattern that's been followed.

File: client/src/modules/reactcomponents.js (hunk 4 above)
Comment (rauenzi) [reply client/src/modules/reactcomponents.js]: This is why I originally had Filters in Utils since I knew it would be reused.

File: client/src/modules/reactcomponents.js (hunk 7 above)
Comment (rauenzi) [reply client/src/modules/reactcomponents.js]: From a quick look it feels like there's a lot of the same interfaces with several new ones. I do agree some refactoring is needed however as there are a couple oddities. But like I said I've only taken a quick look. The reason for this quick implementation over your already done implementation was what we discussed before in keeping track of the patches and not having one cancel another or restore another.

Documentation should come later once the correct functionality is confirmed, which I'm not sure we have sufficiently done.

File: client/src/modules/patcher.js (hunk 2 above)
Comment (Jiiks) [reply client/src/modules/patcher.js]: Temporary

File: client/src/modules/patcher.js (hunk 3 above)
Comment (Jiiks) [reply client/src/modules/patcher.js]: Same here

File: client/src/modules/reactcomponents.js (hunk 4 above)
Comment (Jiiks) [reply client/src/modules/reactcomponents.js]: Didn't see that

File: client/src/modules/reactcomponents.js (hunk 5 above)
Comment (Jiiks) [reply client/src/modules/reactcomponents.js]: Temporary

File: client/src/modules/patcher.js (hunk 6 above)
Comment (Jiiks) [reply client/src/modules/patcher.js]: Then stop thank you

File: client/src/modules/reactcomponents.js (hunk 7 above)
Comment (Jiiks) [reply client/src/modules/reactcomponents.js]: okay bye
